server/client.go

```diff
@@ -814,9 +874,51 @@ func (c *client) processMsg(msg []byte) {
 	if c.trace {
 		c.traceMsg(msg)
 	}
+
+	// Disallow publish to _SYS.>, these are reserved for internals.
+	if c.pa.subject[0] == '_' && len(c.pa.subject) > 4 &&
+		c.pa.subject[1] == 'S' && c.pa.subject[2] == 'Y' &&
+		c.pa.subject[3] == 'S' && c.pa.subject[4] == '.' {
+		c.pubPermissionViolation(c.pa.subject)
+		return
+	}
+
+	// Check if published subject is allowed if we have permissions in place.
+	if c.perms != nil && c.perms.pub != nil {
+		allowed, ok := c.perms.pcache[string(c.pa.subject)]
+		if ok && !allowed {
+			c.pubPermissionViolation(c.pa.subject)
+			return
+		}
+		if !ok {
+			r := c.perms.pub.Match(string(c.pa.subject))
+			if len(r.psubs) == 0 {
+				c.pubPermissionViolation(c.pa.subject)
+				c.perms.pcache[string(c.pa.subject)] = false
```

Comment:
Yes, but we return here. If a publisher publishes on foo.1, foo.2, foo.3, etc... or inboxes that are all unique you may have a case where the cache grows bigger than the set limit. Only after a positive match will you go to the cache pruning.